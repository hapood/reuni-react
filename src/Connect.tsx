import * as React from "react";
import * as PropTypes from "prop-types";
import { StoreObserver, Reuni } from "reuni";
import { shallowEqual } from "./utils";

export type Props<P> = {
  Component: React.ComponentType<P>;
  childProps: P;
  storeObserver: StoreObserver;
  nodeName?: string | null | undefined;
  thread: symbol;
  id?: string | null | undefined;
};

export type State = { childProps: any; isValid: boolean; entityDict: any };

export default class Connect extends React.Component<Props<any>, State> {
  static contextTypes = {
    reuni: PropTypes.any
  };

  static childContextTypes = {
    reuni: PropTypes.object.isRequired,
    reuniNode: PropTypes.string.isRequired
  };

  constructor(props: Props<any>) {
    super(props);
    this.state = { childProps: {}, isValid: false, entityDict: null };
  }

  componentWillMount() {
    let {
      reuni,
      reuniNode
    }: { reuni: Reuni; reuniNode: string } = this.context;
    let { thread, nodeName, id, storeObserver } = this.props;
    reuni
      .mountNode({
        id,
        thread,
        name: nodeName,
        parentId: reuniNode
      })
      .observe(storeObserver, (isValid, entityDict) => {
        if (isValid !== false) {
          this.setState({
            isValid: true,
            entityDict
          });
        } else {
          this.setState({
            isValid: false
          });
        }
      });
  }

  shouldComponentUpdate(
    nextProps: Props<any>,
    nextState: State,
    nextContext: any
  ) {
    if (this.state !== nextState) {
      return true;
    }
    if (shallowEqual(this.props, nextProps) !== false) {
      return true;
    }
    if (shallowEqual(this.context, nextContext.context) !== false) {
      return true;
    }
    return false;
  }

  render() {
    let { isValid, entityDict } = this.state;
    if (isValid !== true) return null;
    let { Component, childProps } = this.props;
    return <Component props={Object.assign({}, entityDict, childProps)} />;
  }
}
