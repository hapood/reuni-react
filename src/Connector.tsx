import * as React from "react";
import * as PropTypes from "prop-types";
import { StoreObserver, Reuni, NodeAPI } from "reuni";
import { shallowEqual } from "./utils";

export type Props<P extends PP, PP extends string> = {
  Component: React.ComponentType<P>;
  childProps: P;
  storeObserver: StoreObserver<PP>;
  nodeName?: string | null | undefined;
  thread: symbol;
  stores: [string, new () => any, StoreObserver<any> | null | undefined][];
  id?: string | null | undefined;
};

export type State = {
  childProps: any;
  isValid: boolean;
  entityDict: any;
  reuni: Reuni | null | undefined;
  nodeId: string | null | undefined;
};

export type Context = {
  reuniInfo:
    | {
        isValid: boolean;
        reuni: Reuni;
        parentId: string;
      }
    | null
    | undefined;
};

export default class Connector extends React.Component<Props<any, any>, State> {
  static contextTypes = {
    reuniInfo: PropTypes.any,
    router: PropTypes.any
  };

  static childContextTypes = {
    reuniInfo: PropTypes.any
  };

  constructor(props: Props<any, any>) {
    super(props);
    this.state = {
      childProps: {},
      isValid: false,
      entityDict: null,
      reuni: null,
      nodeId: null
    };
  }

  componentWillMount() {
    let { reuniInfo }: Context = this.context;
    if (reuniInfo != null) {
      let { isValid, reuni, parentId } = reuniInfo;
      if (isValid !== false) {
        this.mountReuniNode(this.props, reuni, parentId);
      }
    } else {
      throw new Error("Node must under Provider Component.");
    }
  }

  mountReuniNode = (
    props: Props<any, any>,
    reuni: Reuni,
    parentId: string | null | undefined
  ) => {
    let { thread, nodeName, id, storeObserver, stores } = props;
    let node = reuni.mountNode({
      id,
      thread,
      name: nodeName,
      parentId
    });
    this.setState({ nodeId: node.getId() }, () => {
      stores.forEach(([storeName, Store, storeOb]) => {
        node.addStore(storeName, Store, storeOb);
      });
      node.observe(storeObserver, (isValid, entityDict) => {
        if (isValid !== false) {
          this.setState({
            isValid: true,
            reuni,
            entityDict
          });
        } else {
          this.setState({
            isValid: false
          });
        }
      });
    });
  };

  componentWillReceiveProps(nextProps: Props<any, any>, nextContext: any) {
    let { id, stores, thread, storeObserver, nodeName, Component } = this.props;
    if (
      storeObserver !== nextProps.storeObserver ||
      stores !== nextProps.stores ||
      thread !== nextProps.thread ||
      id !== nextProps.id ||
      nodeName !== nextProps.nodeName
    ) {
      let { reuniInfo }: Context = nextContext;
      if (reuniInfo != null) {
        let { isValid, reuni, parentId } = reuniInfo;
        if (isValid !== false) {
          let { nodeId } = this.state;
          if (nodeId != null) {
            reuni.unmoutNode(nodeId);
            this.setState({ isValid: false, nodeId: null }, () => {
              this.mountReuniNode(nextProps, reuni, parentId);
            });
          }
        }
      } else {
        this.setState({
          isValid: false
        });
      }
    }
  }

  componentWillUnmount() {
    let { reuniInfo }: Context = this.context;
    let { reuni }: { reuni: Reuni } = reuniInfo as any;
    if (this.state.nodeId != null) {
      reuni.unmoutNode(this.state.nodeId);
    }
  }

  getChildContext() {
    return {
      reuniInfo: {
        isValid: this.state.isValid,
        reuni: this.state.reuni,
        parentId: this.state.nodeId
      }
    };
  }

  shouldComponentUpdate(
    nextProps: Props<any, any>,
    nextState: State,
    nextContext: any
  ) {
    if (this.state !== nextState) {
      return true;
    }
    if (shallowEqual(this.props, nextProps) !== false) {
      return true;
    }
    if (shallowEqual(this.context, nextContext) !== false) {
      return true;
    }
    if (
      this.context.router != null &&
      this.context.router.route.location !== nextContext.router.route.location
    ) {
      return true;
    }
    return false;
  }

  render() {
    let { isValid, entityDict } = this.state;
    if (isValid !== true) return null;
    let { Component, childProps } = this.props;
    return <Component {...Object.assign({}, entityDict, childProps)} />;
  }
}
