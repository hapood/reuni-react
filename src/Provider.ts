import * as React from "react";
import * as PropTypes from "prop-types";
import { Reuni, createReuni } from "reuni";

export type Props = {};

export default class Provider extends React.Component<Props, { reuni: Reuni }> {
  static contextTypes = {
    reuniInfo: PropTypes.any
  };

  static childContextTypes = {
    reuniInfo: PropTypes.any
  };

  constructor(props: Props) {
    super(props);
    this.state = { reuni: createReuni() };
  }

  render() {
    return React.Children.only(this.props.children);
  }

  getChildContext() {
    return {
      reuniInfo: {
        reuni: this.state.reuni,
        isValid: true,
        parentId: null
      }
    };
  }
}
