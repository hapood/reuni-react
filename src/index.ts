export { storeObserver, value, task, store, delay } from "reuni";
export { default as Provider } from "./Provider";
export { default as createMount } from "./createMount";
export { default as Connector } from "./Connector";
