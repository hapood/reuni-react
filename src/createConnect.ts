import { StoreObserver } from "reuni";
import * as React from "react";
import Connect from "./Connect";

export default function createConnect(thread: symbol = Symbol("thread")) {
  return <P>(
    Component: React.ComponentType<P>,
    storeObserver: StoreObserver,
    nodeName?: string
  ) => {
    (props: P) =>
      React.createElement(Connect, {
        Component,
        childProps: props,
        storeObserver,
        nodeName,
        thread
      });
  };
}
