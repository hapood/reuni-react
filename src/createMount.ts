import { StoreObserver } from "reuni";
import * as React from "react";
import Connector from "./Connector";
import { Omit } from "./types";

export class Mount<P extends PP, PP> {
  _Component: React.ComponentType<P>;
  _nodeName: string | null | undefined;
  _stores: [string, new () => any, StoreObserver<any> | null | undefined][];
  _storeObserver: StoreObserver<any>;
  _thread: symbol;

  constructor(
    Component: React.ComponentType<P>,
    storeObserver: StoreObserver<PP>,
    thread: symbol,
    nodeName?: string
  ) {
    this._Component = Component;
    this._nodeName = nodeName;
    this._stores = [];
    this._storeObserver = storeObserver;
    this._thread = thread;
  }

  addStore(
    storeName: string,
    RawStore: new () => any,
    storeOb?: StoreObserver<any>
  ) {
    this._stores.push([storeName, RawStore, storeOb]);
    return this;
  }

  toHOC() {
    return (props: Omit<P, keyof PP>) =>
      React.createElement(Connector, {
        Component: this._Component,
        childProps: props,
        storeObserver: this._storeObserver,
        nodeName: this._nodeName,
        stores: this._stores,
        thread: this._thread
      });
  }
}

export default function createMount(thread: symbol = Symbol("thread")) {
  return <P extends PP, PP>(
    Component: React.ComponentType<P>,
    storeObserver: StoreObserver<PP>,
    nodeName?: string
  ) => new Mount<P, PP>(Component, storeObserver, thread, nodeName);
}
