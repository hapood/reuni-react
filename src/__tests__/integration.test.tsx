import * as React from "react";
import * as renderer from "react-test-renderer";
import { storeObserver, Provider, createMount } from "src";
import TestStore from "./TestStore";
import TestComponent from "./TestComponent";

it("works with store observing", done => {
  let mount = createMount();
  let TestHOC = mount(
    TestComponent,
    storeObserver(({ testStore }) => ({
      testStore
    }))
  )
    .addStore("testStore", TestStore)
    .toHOC();
  let component = renderer.create(
    <Provider>
      <TestHOC
        onCntAdded={() => {
          let tree = component.toJSON();
          expect(tree).toMatchSnapshot();
          done();
        }}
      />
    </Provider>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
