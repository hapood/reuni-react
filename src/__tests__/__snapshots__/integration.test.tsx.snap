// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`works with store observing 1`] = `
<div>
  0
</div>
`;

exports[`works with store observing 2`] = `
<div>
  10
</div>
`;
